impl_core/src/main/java/io/opencensus/implcore/internal/ShutdownHookManager.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright 2018, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.implcore.internal;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Objects;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * An utility class to help with registering shutdown {@link Hook}s to the runtime.
+ *
+ * <p>A {@link Hook} can have its own name and priority. The execution order among registered hooks
+ * will be determined by their priorities first (descending order) and then registration time
+ * (ascending order). That is to say, higher priority {@code Hook} will be executed first, and
+ * earlier registered {@code Hook} will be executed first given the same priority.
+ *
+ * <p>Caveat: 1. GAE Java 7 environment is not supported for now. 2. User can still use other method
+ * to register a shutdown hook, e.g. the native {@link Runtime#addShutdownHook}. The execution order
+ * of registered {@link Hook}s and other hooks depends on the JVM implementation and thus can not be
+ * guaranteed, which might sometimes leads to unwanted behaviors. An example is that {@code
+ * java.util.logging.Logger} has its own shutdown hook to reset all handlers, and once invoked, it
+ * will prevent all future logging in other shutdown hooks.
+ */
+@ThreadSafe
+public final class ShutdownHookManager {
+
+  private static final Logger logger = Logger.getLogger(ShutdownHookManager.class.getName());
+
+  // The singlelon instance.
+  private static final ShutdownHookManager instance = new ShutdownHookManager();
```

Comment:
Should this be guarded by `monitor` as well?

Reply:
Yes. Thanks for pointing out that.

Reply:
On a second thought, in this case the instance is initialized when the class is loaded so I think it should be fine here.